### ISSUE TITLE
3dfx support

### ISSUE THREAD

**Issue by Zapeth** [dosbox-staging/dosbox-staging#339]
I'm wondering if there is any interest/plans to integrate the 3dfx emulation patch for dosbox (latest one seems to be [here](https://www.vogons.org/viewtopic.php?t=41853))?

I actually already got it working with dosbox-staging (only tested with minGW compiler and Tomb Raider game), though there are some open points (see https://github.com/Zapeth/dosbox-staging/commit/609234ebdb26602d878851804ada2a8f04d03b15).

Since I don't have any immediate references that I could use to address these points so I'll probably just leave it at that state for the forseeable future, but if you think this is (close to) a mergable state I'd also be open for feedback.

**Comment by kcgen** [comment]
Wow - fantastic work!

I like how you first layered in the new files (like `voodoo.cpp`), and then in the next commit integrated it with the existing code.

3dfx support is one of the largest technical hole missing from DOSBox; it represents late-stage capabilities similar to how Hercules and CGA graphics represent early-stage. So technically, I see this is a highly desirable.

I would love to play Voodoo and Glide games in dosbox-staging - however it's an area I've stayed away from simply because many patches were Window-only; and it was quite a maze to navigate.  

Do you know if this patch is Linux/MacOS compatible? Are there pointers you can give regarding how we would test this? (ie: dropping in various shim-files or glide.dll's into a given game's directory?)

**Comment by Zapeth** [comment]
I only recently started to explore this whole endeavor myself, so I'm by no means an expert on the matter.

There are various patches posted in the referenced vogons thread that might address (or even fix) specific issues, but I went with the original to be on the safe side.

Consequently I don't know if this works on Linux and/or MacOS (one of those other patches supposedly fixes some MacOS issues), but at least my changes shouldn't cause problems in that direction.

Something else worth mentioning is that I only checked conflicting SDL/OpenGL code when integrating the patch, so this assumes that the rest of the library-relevant code still works the way it did upstream.

As far as I know you need the same game files that you'd normally use with original hardware in order to test this.

In my case I only have Tomb Raider for testing, and here it seems that you need to include the according glide2x.ovl file in the game directory in order for it to work (in addition to applying the 3dfx patch if you haven't already).

Note that you can't just take any glide2x.ovl, it needs to be game specific, otherwise it might not work or just default to software rendering (I only got it to work with the glide2x_emu.ovl from the gog version which I think is also posted somewhere on the vogons forum).

**Comment by dreamer** [comment]
@Zapeth thank you for working on this :)

As @kcgen mentioned - yes, we are very interested, I will edit the title and put this straight into the backlog.

This change is going to be… simply big, and I am not a fan of dumping large amounts of code in tree at once - I hope we'll be able to split it into smaller, understandable chunks - so the knowledge about how this patch *actually works* will reach other developers (this is how we dealt with CD-DA and SDL2 - both successfully, and we learned a lot in the process). Perhaps it'll turn out to be very time-consuming in practice, so we'll need to find an arrangement, that will allow this work to move forward without compromising code quality in master branch.

If you plan to keep working on it, I can give you push access to this repo - you won't be able to push to master branch, but your private branches will run against our CI on every push, thus speeding up development. You can also enable GitHub Actions in your fork, with similar benefits.

---
### About the patch itself

Last time I investigated patches about hardware acceleration, it was very early in dosbox-staging life (September or October 2019). I don't remember which version exactly I was testing, but I needed to mess with `*.ovl` files, so I guess it was a version with a Glide patch.

So my test results might be completely irrelevant to this work, but *back then* I was testing following games (on Linux only, it was *before* we moved to SDL2):

- Tomb Raider - it "worked"; game was perfectly playable and worked fine. But there were issues with emulator window, fullscreen, input, and no cutscenes at all
- Carmageddon - it ranged from "not working at all" to "barely starts" - e.g. I was able to start race, but there was no text at all, no menus (only black screens)
- Descent (I don't remember 1 or 2) - I couldn't make it run

### Related work

Since then, some versions of these patches got imported twice to this repo, but I did not look too much in details around them:

1. A version from ECE, the patch is imported to branch `po/ece-r4305-1`: 594e3209e38e5bcf34f772df9164c7f1f793fafb
2. A dumped version of DOSBox source code distributed with Carmageddon - one of DOSBox maintainers worked on this version - it is 0.74 with one of the community patches applied on top (and perhaps some additional work); imported to branch `steam/carmageddon`: 1cedad5a897b33c33a1b8fe4c1729aeedb65f6cd

### TL;DR; for new people finding this thread

We can't just slap this code in master branch, we need to understand how it works, when it works, what are the limitations, review it properly, etc. I think splitting this work into small, reviewable pull requests is necessary.

**Comment by dreamer** [comment]
@Zapeth do you want me to assign you? (it will be a clear signal for the community about who's working on this). There's no obligation involved, of course - if you won't have time to work on this in the future, someone else could pick up your work.

**Comment by dreamer** [comment]
It does not compile on Linux OOTB; I slapped compilation fixes and pushed it to branch: `po/3dfx-emu-1`.

Overall results: :(

### Carmageddon (Steam release, seems the same as GOG release)
Used following settings (my usual ones with `pci` section added, as introduced by this patch):
```
[dosbox]
machine = svga_s3
memsize = 32

[pci]
voodoo = opengl
voodoomem = standard

[autoexec]
mount C ".." -freesize 100000
imgmount D "../CARMA/boxtron.cue" -t cdrom -fs iso
C:
cd CARMA
3dfx.exe
```
Does not work - the game starts in 640x480 software mode, work like it did before.

I *think* some warning about Glide flashes a moment before starting the game, but the game clears the screen on exit :(

When starting another binary distributed with the game instead: `voodo2c.exe` - same result (perhaps message is a different one, but it flashes too fast).

### Tomb Raider (Steam release, seems the same as GOG release)
Also added `pci` section, running `TOMB.EXE` instead of usual `TOMBno3Dfx.EXE`.

Does not work:
```
Fatal error: unable to load DLL.
```
### Descent 1, 2 (Steam, same as GOG)

Turns out games are distributed without 3dfx patches.

### Shadow Warrior (freeware on Steam)

Haven't tried it, but [this](http://www.r-t-c-m.com/knowledge-base/documents-online/setup-dosboxemu-3dfx-sw.html) suggests it requires Glide wrapper.

**Comment by shermp** [comment]
I thought I'd chime in with the fact that PCem has very usable Voodoo1/2 software emulation, if anyone wanted to take a look at it. Have no idea how difficult it might be to integrate, and it's plain C, which some may consider a downside. I'm afraid hardware emulation goes waaaay over my head.

Performance wise, it's... ok? I tried the Descent 2 (D2VOODOO patch), and when PCem wasn't crawling to a slideshow (seemed to happen for particular explosions/pickups, and don't know whether it was CPU or V2 slowdown), performance for a single V2 (recompiled) was around 60fps. The interpreter gave me around 30fps.

It also runs the intro sequence of Unreal Gold at around 15-30fps, (640x480, high, Pentium MMX 200), which I thought would probably be quite playable, especially if one turned down the resolution or details.

Note, my host CPU is an i5 4690K, so not exactly a modern powerhouse...

**Comment by Zapeth** [comment]
@dreamer Thanks for the extensive answer, I'll try to address the most important points:
> This change is going to be… simply big, and I am not a fan of dumping large amounts of code in tree at once - I hope we'll be able to split it into smaller, understandable chunks - so the knowledge about how this patch _actually works_ will reach other developers (this is how we dealt with CD-DA and SDL2 - both successfully, and we learned a lot in the process)

The original patch is pretty much self-contained and only adds code to original files for the most part, so I'm not sure if it would be useful to try to split it up in smaller chunks.

On the other hand, my changes to get the patch working with dosbox-staging on minGW are not that many, but I can split them up in single commits on another branch if you want.

> If you plan to keep working on it, I can give you push access to this repo - you won't be able to push to master branch, but your private branches will run against our CI on every push, thus speeding up development.

I don't think that will be necessary right now, but if the nobody else wants to work on it in the coming months then I might get back to you (hopefully I'll have more free time available then than right now).

> Carmageddon testing

Do the same game files work with dosbox-ece? (you'll know when some `VOODOO:` messages show up in the console)

>Tom Raider testing

`Fatal error: unable to load DLL.` indicates that the game can't find/load the glide2x.ovl (needs to be in the same directory as the game executable).
Alternatively you can try the [voodoo rush patch](https://tombraiders.net/stella/files/gamefiles/TR1files/TR1_rush_patch.zip) which should work out of the box.

While testing that game I noticed a small visual glitch, namely a white outline sporadically showing up around Lara's left pistol and around the pistol's muzzle flash sprites.
However this also shows up in other dosbox versions that include the patch, so this is most likely an unresolved issue.
______________________
I also got around to test Screamer2 with the following results:
- it only works with the voodoo 1 patch of the game, newer cards are apparently not emulated by the 3dfx patch (something to keep in mind for testing with other games)
- when starting the game it only shows a black screen, but a few seconds into the intro it renders the game (albeit with some flickering on the ingame text).
  - Note that the game's CD needs to be mounted for the intro to show up, otherwise it will only show an "Insert CD" screen which is also not rendered.
  - This issue is not present with other dosbox versions that feature the 3dfx patch, so either those versions contain some fix for this, or its a dosbox-staging specific issue (either due to improper patch integration or other differences that come with SDL2)

@shermp thats certainly something to keep in mind (as well as checking the most recent voodoo-relevant code in mame, since this patch was using that as a reference).
However I think the most immediate points should be resolved first, which are:
- getting a testable state (for all platforms)
- fixing remaining issues that are not present in other dosbox versions that feature this patch (like the Screamer2 issue I mentioned)
- improve integration with SDL2 (properly deal with other dosbox output modes, improve fullscreen rendering, add option(s) that fit the rendering resolution to dosbox output and/or just override it)

**Comment by MartyShepard** [comment]
Hi,
A long time ago I integrated the 3DFX patch from Kekko into my Dosbox Fork and adapted it to SDL2 (dynamic resolution up to 8k) and have worked on it ever since.
Kekko definitely likes the changes.

https://www.vogons.org/viewtopic.php?p=837590#p837590

Attention! The patch is based on parts of Mame anno ~ 2010 and 2013. Since then, Mame (Author Aaron Giles) has done a lot in the area of voodoo. Actually you would have to adapt Mame's source (Emu and OpenGl) to the patch.

**Comment by kcgen** [comment]
@MartyShepard , forgive my lack of context - I see last month **kjliew** updated his combo-patch (attached) of both **Kekko**'s 3dfx patch with Gulikoza's glide pass-through patch, which **yesterplay80** nicely described as offering ...

> "even more ways of emulating a Voodoo card now. While kekkos patch includes an internal wrapper that “translates” 3Dfx calls to OpenGL ones, gulikozas patch passes 3Dfx calls through to an external wrapper like nglide or dgvoodoo2, offering more options to adjust the image to your liking.. Please note that, depending on the method you prefer, you might need different OVL files in DOSBox, please [refer to this post](https://www.vogons.org/viewtopic.php?f=32&t=52414&p=791672#p790831) in the vogons forum for more information."

Do you know if **kjliew**'s patch is based on your improved **Kekko** 3dfx patch?  

[v1emu_glide2h_r4335.patch.gz](https://github.com/dosbox-staging/dosbox-staging/files/4569322/v1emu_glide2h_r4335.patch.gz)

**Comment by MartyShepard** [comment]
Hello kcgen :)
I haven't tried the patch, but I think we shouldn't be dependent on dlls and wrappers. In the end, this always led to users not knowing exactly what really worked. It's also about fully integrating the hardware and modifying it to run at full speed (640x480 doesn't need static). Either wrapper or fully integrated. but not both.

to kjliew's patch
I can't tell you now. do you have a direct link for the patch.

The best approach should be carefully considered. This is sometimes the problem with Dosbox. Especially when it gets modern. Everything is rinsed in. I would rather prefer 3DFX and the following models.

The bottom line is to maintain the machnik. but people want to play iti 3D in best off. ;)
It should be old but also new;)

**Comment by kcgen** [comment]
> this always led to users not knowing exactly what really worked

Great point; my own reluctance to dig into the DOSbox-Voodoo side of gaming (on Linux no less) is a perfect example of this. 

> kjliew's patch: do you have a direct link

Yes, it's the 'v1emu_glide2h_r4335.patch.gz' attached file (at the bottom) of my post.

I just want to make sure that we're starting off from the best baseline the community has already progressed.  From the sounds of it, that's probably your improved version of Kekko's patch.

Regarding the two mechanisms in kjliew's combo: I see the problem somewhat like MIDI.  DOSBox itself isn't a MIDI player, instead it passes MIDI calls out to various players such as MT32, Timidity, BassMidi, FluidSynth, Windows-native, macOS native.. and so on.   In the next revision of dosbox-staging we want to internally interface and build-in MT32 and MIDI playback; however the ownership and maintenance of (whatever we link-up-with) will still fall to those respective projects.

In the same way, outputting 3dfx calls to an external interpreter (similar to the MIDI player), can offload that work for us.  Those external 3dfx-to-host translators can implemented OS-specific enhancements and do the work of "keeping up with updates".  The field is already quite fragmented with OpenGL, OpenGLES, Vulkan, DirectX, WebGPU, and Metal.

On the other hand, I also appreciate the ability to become stand-alone and self-contained. If we can translate from 3dfx to rasterized SDL2 calls, then we can allow the SDL2 project to abstract using from having to (ultimately) write Metal and Vulcan API calls (when OpenGL is eventually replaced by these). 

I think the second approach is probably the right way.  I'm confident will have keen 3D developers who will step up to those translation effort when the time comes.

**Comment by dreamer** [comment]
*Sorry about not being too active in this discussion, I am still occupied with preparing 0.75.0 final release.*

I agree with @MartyShepard - we should not depend on dlls and wrappers - I think for the time being we should focus on cleaning up code started by **Kekko**; more about this in section "next tasks" below.

@Zapeth 
> Generally speaking I think we should differentiate between dosbox-staging specific issues and (unresolved) patch issues.

Just to be clear: all game referenced until now (with exception of Screamer, which I've never tested myself) work *perfectly fine* in dosbox-staging in software modes. Tomb Raider, Descent 1, and Descent 2 hit maximum framerates allowed by their respective VGA/SVGA emulation modes (50 or 70FPS). Carmageddon works perfect in software in low resolution mode, and is *playable* (not super, but playable) in software 640x480 mode. There might be some small issues (e.g. we have problem with Tomb Raider on ARM), but nothing is critically broken (as far as I can tell, but I test mostly on Linux x86_64).

Anyway, thank you very much for link for that 3dfx patch to TR1! 😊 

I can finally confirm - it works, on Linux, no messing with *.ovl patches required, I didn't need to compile openglide manually, it worked. Performance is good, cutscenes work, no input issues. This is **the first time** I have seen 3dfx emulation working *acceptably* in any dosbox on Linux (there are bugs you mentioned earlier).

This gives me some confidence, that this is the proper development path to follow. And @MartyShepard confirmation and screenshots from other games are good indicator as well.

### Next tasks

The patch, in it's current form is *too big*, no - it's not well contained. We need to split it into smaller tasks, identify which parts need only slight adaptations, which ones need newer code from MAME project, and what parts are completely missing or need to be rewritten from scratch.

So far I had only brief look at the code, but I already identified 4 actionable tasks, that will bring this feature closer to integration:

#### 1. Avoid duplicating OpenGL initialization

When original patch by Kekko was written, dosbox code did not query for OpenGL extensions at all - that's why `src/hardware/voodoo_vogl.cpp`, `src/hardware/voodoo_vogl.h`, and `src/hardware/voodoo_types.h` were written.

Upstream project decided to use and modify some patch by NY00123 to introduce GLSL shaders - and also introduced very similar code in SVN trunk (which we inherited) - it's in `src/gui/sdlmain.cpp`, parts of this code @Zapeth modified to fix compilation problems.

We can't have 2 places in code, that are responsible for the same thing - it will cause serious problems in the future. The first task is to elevate code implemented in `src/hardware/voodoo_vogl.cpp` up, so it can be reused for GLSL shader initialization and eventually could be removed from original Kekko's patch.

My suggestion would be to place this code in files: `include/dosbox_gl.h` and introduce new code module: `src/gl/` with file `src/gl/gl.cpp` inside - this code should take over opengl initialization that lives in `src/gui/sdlmain.cpp` right now… although person who will implement this might change my mind about placing or naming of these files :)

*This is already big task, but it's a prerequisite for future development work*

#### 2. Provide buildsystem switch to enable/disable the feature

We won't finish this work before next stable release 0.76.0, and I don't want to keep pushing stable releases away forever, because one feature is not finished. We need a way to disable this feature from stable release builds, as to prevent users from deciding the project is broken, reporting issues for stuff we know does not work, etc, etc.

In other words: we need a new configure switch, e.g.: `--disable-3dfx-emu`

#### 3. Clear-out legal and licensing issues with MAME project

The patch includes code merged in from MAME project. MAME is GPL-licensed, so theoretically we could just merge this in… but the disclaimer put in these files prevent us from doing so (I don't know if this code is GPL).

We need to:
- clear this issue with original author (I am especially worried about 3rd point from license in those files)
- establish what license is being really used in there, so we can mark it with correct SPDX identifier
- put in copyright date based on the version of code from MAME project, because existing "Copyright" line without dates is legally doubtful
- make sure our other legal obligations are fulfiled (e.g. as distributing license terms for this code in with our CI builds)

*This is not as scary as it sounds - I am 100% sure we'll be able to resolve this with MAME project without problems - but we need to contact the MAME maintainers and original author; I am just being prudent maintainer here.*

*Stupid problems like this happen when people don't use standard open-source licenses and put in their own crafted ones in code instead.*


#### 4. Start wiki page where we can share info on 3dfx-enabled games

We need a place to communicate about list of games with 3dfx support, what patches are required to make the game work, what testing was done already, critical bugs in those games, etc.

This information will be essential for testing and future development.

Let's focus only on DOS games for now, and leave out Windows 95 games - there should be only few titles (less than 20?). 

*I am sure info like this is somewhere on Vogons already, but I do not have access to Vogons forum. Keeping info locked on that forum slows us down.*

Our project wiki on GitHub is open for everyone to contribute :)

**Comment by dreamer** [comment]
> > this always led to users not knowing exactly what really worked
> 
> Great point; my own reluctance to dig into the DOSbox-Voodoo side of gaming (on Linux no less) is a perfect example of this.

same here

> I just want to make sure that we're starting off from the best baseline the community has already progressed. From the sounds of it, that's probably your improved version of Kekko's patch.

I think the mega patch of combining all of 3dfx work was a huge overreach. *There was no other way to keep this work alive as an out-of-tree patch on Vogons forum, though.*

> I think this second "embedded & standalone" approach is probably the right way, even if we're still translating 3dfx calls to OpenGL. I'm confident we will have keen 3D developers who will step up to those translation efforts (to Metal and Vulcan) when the time comes.

Right now the patch provides 2 backends for voodoo emulation: software and opengl; software has terrible performance though (so terrible in fact, that I feel like we should not make it available to users in release builds and keep it only as an option for debug builds).

The really nice thing about this is: the code is already abstracted, so we'll be able to provide new backends in the future - also, missing parts of voodoo emulation are clearly marked in the code (!).

Maybe we can even implement fast SIMD-powered software backend.

I don't want to touch Metal with a ten-foot pole - I hope Vulkan+MoltenVK will be a better solution (but it's not directly relevant to this work yet). However, if we'll find a maintainer willing to work on Metal backends… (right now opengl backend in voodoo patch is 1909 LOC).

**Comment by Zapeth** [comment]
In regards to discussion about Voodoo Emulation vs Glide Wrapper there are a few pros and cons for each.

Voodoo Emulation:
- should theoretically support every game that uses 3dfx in DOS (for this reason alone I see this as the superior choice), its just currently not the case because the patch hasn't been developed far enough to emulate more hardware
- doesn't require external dependencies (only whatever the game needs to use 3dfx, like an .ovl file), this makes it somewhat easier for the end user but also means that more code needs to be maintained in one place (as opposed to offload it to an wrapper dll)

Glide Wrapper:
- can only support DOS games that require an .ovl file for 3dfx (for example the Tomb Raider patch I linked above wouldn't work with this method)
- offloads the "complex" part of this feature to the wrapper, leaving the choice up to the user which one to use (good for power users, bad for people who just want to play) and less work maintaining the dosbox code parts (only need to ensure that the wrapper is properly integrated)
- most existing wrappers offer pretty good 3dfx emulation (superior to current voodoo emulation patch), however as far as I can tell there are no active _open source_ glide wrappers (openglide exists but seems to be pretty much on life support these days)

It is possible to integrate both methods which currently offers the best game compatibility but will probably also confuse the hell out of most end users if problems occur, and put more work on developers who need to maintain both features.

@dreamer 
> Just to be clear: all game referenced until now (with exception of Screamer, which I've never tested myself) work perfectly fine in dosbox-staging in software modes.

I was actually just referring to the 3dfx functionality part of these games, unless you meant that they worked fine with voodoo emulation being set to `software` and `VOODOO:` log entries showing up in console when starting the game?

> Right now the patch provides 2 backends for voodoo emulation: software and opengl; software has terrible performance though (so terrible in fact, that I feel like we should not make it available to users in release builds and keep it only as an option for debug builds).

Actually I think it would be easier to always tie the voodoo output method (currently configurable in the config) to the configured dosbox output method, that way you dont have to do stuff like [this](https://github.com/Zapeth/dosbox-staging/commit/609234ebdb26602d878851804ada2a8f04d03b15#diff-35ee3fc19dedf0fcf3ff2f595d40a611R1586-R1589) since the backends would always match (this is just catching the erroneous case, but proper backend switching would be a lot more tedious and error-prone).

**Comment by dreamer** [comment]
> I was actually just referring to the 3dfx functionality part of these games (…)

Oh, then I simply misunderstood - sorry about that :)

>  * most existing wrappers offer pretty good 3dfx emulation (superior to current voodoo emulation patch), however as far as I can tell there are no active _open source_ glide wrappers (openglide exists but seems to be pretty much on life support these days)

This reason alone is a deal-breaker here. If one of the existing wrappers changed to open-source, GPL-compatible license (and got ported to other operating systems), then we might consider it… probably not going to happen though. Right now, we can't cooperate with those projects, and can't properly support them in dosbox-staging.

In last few months I in my spare time I was looking here and there if there are no good alternatives (e.g. maybe Wine had Glide translation layer the same way they have for DirectX - but nope, I also did some digging in Mesa - seems like Mesa's Glide implementation was never finished and it was removed from the code years ago).

I haven't found a single good candidate (but only recently users informed me about PCem and 86box implementations - I haven't investigated them yet).

On the positive side: 3dfx published Glide source code, API references are easy to find, we can always look up or even merge some code from OpenGlide…

> It is possible to integrate both methods which currently offers the best game compatibility but will probably also confuse the hell out of most end users if problems occur, and put more work on 
developers who need to maintain both features.

Thanks for mentioning this - we *do not* want to leave users confused and provide multiple options, each one broken in different scenarios. Going with one implementation will make it easier to test and fix in the long run.


> Actually I think it would be easier to always tie the voodoo output method (currently configurable in the config) to the configured dosbox output method (…)

:+1: Please note you shouldn't worry about `output = surface` too much. Over last 2-3 months this stupid option caused me so much grief during testing… at *no* benefit to the users. It's very likely that we'll drop it completely to make the code cleaner and testing faster. Usage scenarios, that demand software-backend can use `texture_renderer = software` already.

Building upon this idea, we might want to drop `[pci]` section completely, and provide new `machine` option instead, e.g.:
```
[sdl]
output = texture

[dosbox]
machine = svga_voodoo
```
would turn on svga_s3 (?) + voodoo + software backend, while:

```
[sdl]
output = opengl

[dosbox]
machine = svga_voodoo
```
would turn on svga_s3 (?) + voodoo + opengl backend.

Perhaps other option than svga_s3 will work better - the goal is to have 1-2 configurations for testing instead of matrix of (3 outputs x N machines x 2 backeds).

**Comment by BPaden** [comment]
Chiming in to opine that output is conceptually different to the voodoo implementation. The reason to use software emulation is for accuracy and compatibility, but the main reason to use opengl output is for shaders and fast scaling. You may very well want both. I use software emulation for both ps1 and n64 emulation in combination with a shader.

About performance, I do believe that pcem's implementation is a good choice for low level voodoo emulation that runs at full speed on a good cpu. Usually. It's not perfect. In particular, I noticed in Screamer 2 and Carmageddon that the menus run at roughly half speed with the 3dfx patch for some reason, while the actual games ran at full speed.

**Comment by dreamer** [comment]
@BPaden when it comes to output: I think there is no accuracy nor compatibility gain from using a software `output = surface`. Maybe there were reasons 10 years ago, but not any more :). And there's no reason for us to support 2 distinct software modes (with one being special PITA for developers and testers, siphoning development effort away from more important tasks).

About treating output and voodoo backend as the same option - I think it makes sense. We don't want to support mixed scenarios where user selects e.g. `texture_renderer = metal` and voodoo backend selected to opengl. At the same time - when user selects `output = opengl` - there's no reason *not to* select faster opengl as voodoo backend as well.

I agree, that `output` option is not correctly named and might be misleading or unintuitive. It is *already* problematic since someone had an idea to merge API name with scaling hints to create `openglnb` (which was stupid and lazy decision). We stuck to value name `texturepp` (for now) to preserve compatibility for users of pixel-perfect mode from ECE, but literally every developer working on it agreed, that this option should be renamed to something more user-friendly.

 Perhaps, when voodoo support will be good enough for turning it on in release builds, we should rename `output` to something like "`rendering_backend`" (?). What do you think? :)

**Comment by BPaden** [comment]
I agree that `output` is bad experience of poorly documented options that seem redundant or of dubious value. I've used `output = opengl` for years and really don't see any reason to use anything else — except maybe vulkan if that happens. Are these options kept around for compatibility with the amiga or something? But I think the fact that I mentioned software emulation being more accurate and you started talking about `output = surface` just demonstrates that conflating the voodoo implementation with whatever is putting a framebuffer on the screen is just confusing. 😜

I do understand your point. Retroarch had a similar problem for years. If you chose vulkan as the output and the emulation core used opengl, retroarch would simply crash. Everything seems to just work now, but I'm not sure how much work it was to make that happen.

But I wonder if it's true that there is no reason to use LLE. Typically there are big accuracy benefits to emulating the actual 3d hardware accurately in this era. I know there is interest in getting more accurate results than you would see in something like nglide. See https://pcem-emulator.co.uk/phpBB3/viewtopic.php?f=4&t=3028 for an example. Nglide just forces 32-bit color. Real hardware uses 16-bit with dithering, and the the image is run through a dedithering filter on the hardware. Basically there's color banding everywhere on a real voodoo.

While I guess most people want to upscale their old 3d games to 4k so you can count each and every seam in Lara Croft's mesh with perfect clarity, some of us want to keep that chunky blurry mess from the old days around.

But maybe perfectly accurate results can be had with just modern gl+shaders. I understand emulating really esoteric hardware like the n64 RDP or the ps1 is different from a fixed function api like glide. I assume other advantages like d3d and opengl on Windows are also out of scope for dosbox. But what is the situation with ms-dos games statically linking glide? Also does glquake work?

**Comment by DOOMMARINE117** [comment]
interesting post. im suprised people can play windows 95 games, lol. i didn't know those were supported in teh original dosbox? funny, lol. i pretty sure 95, 98, 2000 and xp  have different programs to run with. 

dosgames are inetresting to work with.

**Comment by Kirtai** [comment]
Regarding the MAME licensing, MAME changed their license from a no-commercial-use to [GPL/3BSD](https://docs.mamedev.org/license.html) a while ago. It's possible the code in the patch comes from the pre license change versions. If so, it should be possible to use the code from more up to date releases under the newer licenses.

**Comment by dreamer** [comment]
@Kirtai thank you! That explains those weird license clauses - and it means resolving [task (3)](https://github.com/dosbox-staging/dosbox-staging/issues/339#issuecomment-623049560) should be even easier :)

**Comment by rderooy** [comment]
An updated voodoo patch, with various bug fixes was posted by kekko a week ago.
https://www.vogons.org/viewtopic.php?p=865994#p865994

**Comment by dreamer** [comment]
And right away the patch was imported as dosbox-staging branch (`vogons/voodoo-r4353`): fcc18fface838467f19135d99e5850287f550bb1 ;)

We tried to contact kekko to ask if maybe he won't be interested in working with us to merge hist patch to master, but haven't heard from him yet :(

**Comment by rderooy** [comment]
You may also want to have a look, probably separate from this, at the PCI re-work that was done in dosbox-x.

**Comment by kcgen** [comment]
Kekko mentioned not actively working on patch but is glad to know that we appreciate it and are (or will) work to improve it.  
We should share our improvements in this Vogons thread: https://www.vogons.org/viewtopic.php?f=41&t=41853, as suggested by Kekko.

**Comment by kekko-dos** [comment]
Hi everyone!
I want to thank you all for your efforts. 
I have not worked on this in the last decade and, looking at the patch today, I can tell it needs a serious rewrite ;)
Best option would be starting from the newest mame code, but as you may already know, I can tell you it would be quite a lot of work.
Apart from this, if you have any questions, just shoot, I'll try to recall something about this old code, even if actually I had some troubles in recognizing what I wrote so long time ago.

About using software renderer or host gpu, this mostly depends on what we want to achieve. I knew of most people which just prefer emulation which is perfectly faithful to the original hardware, and for this reason they just code everything in software, even if this means getting 2fps. Personally I prefer getting less accuracy but playable games. On the other hand I always hated dithering :)
It's a matter of tastes, and in my opinion if you want the emulator to be complete  it should provide every possible option (i.e. software renderer with perfect emulation, host gpu acceleration + shaders for dithering, host gpu +32bpp color+ high res).

Thanks again and keep it up!

**Comment by dreamer** [comment]
@kekko-dos Thank you for showing up, we will definitely pick up your offer and ask questions :)

dosbox-staging focuses on ease-of-use - I would prefer if we kept 3dfx emulation backed up by OpenGL by default (OpenGL is already the default "output" we use) with keeping software option hidden for developers only (fewer pointless user-facing options means we can focus on one solution providing the best possible quality instead of several mediocre options). After OpenGL working correctly, we would probably pivot to providing either OpenGL ES or Vulkan backend (seems like every passing week makes Vulkan more compelling to use ;) ).

I plan to start working on this feature after pending tasks in CPU emulation area will be cleared and FluidSynth code will be merged (Fluid is ~85% done at this point - the remaining parts are usability and configuration concerns).

**Until then, anyone can pick it up and start working towards merging it in**; I suggested the first task to be implemented in [this comment](https://github.com/dosbox-staging/dosbox-staging/issues/339#issuecomment-623049560) (section "Next tasks").

Does anyone know the definitive list of DOS games benefitting from 3dfx hardware? (I am not interested in Windows games right now; time for that will be after we'll have majority of DOS games working correctly).

**Comment by BPaden** [comment]
Here is **a** list. I'm not sure how complete it is. http://www.zeus-software.com/forum/viewtopic.php?f=10&t=277&p=1044#p1044

**Comment by Zapeth** [comment]
> Here is **a** list. I'm not sure how complete it is. http://www.zeus-software.com/forum/viewtopic.php?f=10&t=277&p=1044#p1044

There seems to be a more recent version of this list [here](http://www.zeus-software.com/forum/viewtopic.php?f=10&t=672&p=3473#p3473).

There are also a couple of demos that feature 3dfx, though I was only able to execute one or two of them with the patch and the official 3dfx dos driver file, the rest either crashed or wouldn't start (I believe some of them required a voodoo 2 card) -> http://www.pouet.net/lists.php?which=72

**Comment by kcgen** [comment]
See post here: https://github.com/MartyShepard/DOSBox-Optionals/issues/5#issue-691760345 by @stilett0

**Comment by Python-Exoproject** [comment]
For test cases our latest release of eXoDOS (version 5) includes a playlist with all the DOS 3DFX games. 

In eXoDOS we've set them up using ECE, usually in hardware opengl mode as thats the only mode that runs them at a playable frame rate. There are annoying oddities though like hardware mode ignoring any scaling and only using original Voodoo 1 resolution (Voodoo 2 or SLI Voodoo 1 allowed higher even back in the day), and games like JetFighter Full Burn that work fine until the sun comes into view.

**Comment by kcgen** [comment]
@Wengier's picked up this effort under his branch, `ww/voodoo-3dfx`
:rocket: !

**Comment by FeralChild64** [comment]
Just to make sure you know - there is an updated openglide available here: https://github.com/jenshemprich/MacGLide/tree/master/MacGLide/OpenGLide . Still old, but much newer than original openglide.

Although it is MacOS only, it seems only the OpenGL context handling is MacOS-specific.

**Comment by rderooy** [comment]
For DOSBox-X SDL2, I have been suggesting https://github.com/kjliew/qemu-xtra

**Comment by dreamer** [comment]
An update for people watching this thread, but not looking at PRs.

We decided that the best way forward is going with software-only multithreaded version of the patch - it provides good game compatibility, reasonable performance, and a relatively low number of usability issues and other bugs. The initial support for Voodoo was merged in #2410.

The code is below our usual quality standards, but now we can improve it incrementally. ATM the code is hidden behind the feature flag to prevent delaying 0.81.0 release, the plan is to enable it by default once UI/conf will be finalized, code quality will be good enough, and performance will improve (similarly how we did for FluidSynth in the past).

ATM people who want to test it need to compile the code by themselves:

- Linux / macOS / Windows MSYS: `meson setup -Dexperimental_voodoo_support=true build`
- Windows MSVC: manually turn on `C_VOODOO` define in `src/platform/visualc/config.h`

**Comment by FeralChild64** [comment]
I've just noticed two compiler warnings when using GCC 13.1.0. I'm not sure if they are new to this GCC version, or if they were there before:

```
[260/274] Compiling C++ object src/hardware/libhardware.a.p/voodoo.cpp.o
In function 'void raster_fastfill(void*, INT32, const poly_extent*, const UINT16*)',
    inlined from 'void fastfill(voodoo_state*)' at ../src/hardware/voodoo.cpp:4829:19:
../src/hardware/voodoo.cpp:3553:24: warning: 'dithermatrix' may be used uninitialized [-Wmaybe-uninitialized]
 3553 |                 UINT64 expanded = *(UINT64 *)ditherow;
      |                        ^~~~~~~~
../src/hardware/voodoo.cpp: In function 'void fastfill(voodoo_state*)':
../src/hardware/voodoo.cpp:4740:16: note: 'dithermatrix' declared here
 4740 |         UINT16 dithermatrix[16];
      |                ^~~~~~~~~~~~
```

**Comment by dreamer** [comment]
Yep, I have them as well on Fedora and they happen on MSYS2 as well (but not on GCC 12 nor Clang). They are a bit weird - 0-initializing `dithermatrix` does not remove the warning for me.

**Comment by Torinde** [comment]
How many threads is the software emulation using? Up to 4 or up to the max supported by the CPU(s) available?

**Comment by dreamer** [comment]
@Torinde It uses 4 threads.

However, when I limited it to 2 threads for testing purposes (on my 8-core CPU) I noticed a modest performance improvement.

**Comment by Torinde** [comment]
Thanks! One more reason for me to finally install Visual Studio...

Aren't more threads supposed to increase performance (at least until certain quantity)? I'm confused by your comment - do you say that 2 threads is faster than 4 threads?

**Comment by johnnovak** [comment]
> Aren't more threads supposed to increase performance (at least until certain quantity)? I'm confused by your comment - do you say that 2 threads is faster than 4 threads?

Depends entirely on the algorithm, implementation, etc. I can come up with a lots of scenarios where higher parallelism could be actively hurting performance. E.g. copy 1 large, or 2, 3, 4, 5 etc. large files at the same time. The more files = the longer it takes to copy...

**Comment by dreamer** [comment]
> Thanks! One more reason for me to finally install Visual Studio...
> 
> Aren't more threads supposed to increase performance (at least until certain quantity)? I'm confused by your comment - do you say that 2 threads is faster than 4 threads?

Nope, it is called "thread scaling" and, as @johnnovak said - it depends on the algorithm used (and the problem being solved). The perfect scenario is when the problem scales linearly - that is, when you use X threads, then whole problem is solved in 1/X time of single-threaded solution, but it almost never works this way.

In this case we might just be affected by too many context switches, and most likely should improve the way we synchronize the output of worker threads. We had a similar problem when we started work on FluidSynth and after several iterations ended up with simple and efficient solution.

**Comment by dreamer** [comment]
Yes, it stays experimental and will **not** be enabled in our release builds. I'm not even sure if we should mention it in the release notes.

Hopefully, we can shape it up for 0.82.0, but it's not decided yet.

**Comment by Torinde** [comment]
Thanks @dreamer and @johnnovak... after all the "graphics is a parallel workload" and "Voodoo 2 SLI with 4 chips doing texturing and 2 chips doing rendering" I hoped the optimum here is more than 4 threads...

As for release notes - IMHO, no mention if it's not enabled in the release build. On the other hand - I'll be curious to read a blog/issue/discussion post on your "roadmap" plans in general (e.g. the thoughts behind https://github.com/dosbox-staging/dosbox-staging/projects/3, https://github.com/dosbox-staging/dosbox-staging/projects/7)

**Comment by johnnovak** [comment]
By no means should we mention it if it's not ready for prime time. Features hidden behind compile or even secret runtime feature flags are just for us devs.

**Comment by Torinde** [comment]
[Glidos is now open source](https://www.glidos.net/code.html?lang=en) - executing Glide on the host D3D or OpenGL, I assume.

**Comment by johnnovak** [comment]
Closing as Voodoo support is available in the current dev build.

**Comment by GranMinigun** [comment]
Shouldn't this be left open until the feature is actually deemed complete and not merely present?

**Comment by johnnovak** [comment]
> Shouldn't this be left open until the feature is actually deemed complete and not merely present?

Happy to reopen if that helps anything, but not even sure what "complete" means. I mean, Voodoo games are working now. I guess further enhancements will be new tickets? But sure, I can reopen, but as I see it, support has been added, it works.

**Comment by kcgen** [comment]
Voodoo is fully enabled - no longer a built-time flag, [passing CI](https://github.com/dosbox-staging/dosbox-staging/pull/2710/commits), and is working just as well as it does in software mode in ECE, X, and Pure.  

We're just refining a couple bits left prior to release, tracked in this open ticket: https://github.com/dosbox-staging/dosbox-staging/issues/2803
